Source/JavaScriptCore/runtime/WaiterListsManager.h

```diff
@@ -0,0 +1,330 @@
+                return waiter->isAsync() && waiter->getVM() == vm;
+            });
+            bool isOKString = false;
+            notifyWaiterImpl(ptr, WTFMove(waiter), isOKString);
+        }
+    }
+
+    void unregister(VM* vm)
+    {
+        m_registeredVMs.remove(vm);
+    }
+
+    void dump(PrintStream& out) const
+    {
+        out.print("--------------- WaiterListsManager ---------------\n");
+        m_registeredVMs.dump(out);
+        {
+            MutexLocker locker(lock);
+            out.print("\nm_waiterLists (size ", m_waiterLists.size(), "): ");
+            for (auto& entry : m_waiterLists) {
+                out.print("<ptr: ", RawPointer(entry.key), ", ");
+                WaiterList* queue = entry.value.get();
+                RELEASE_ASSERT(queue);
+                {
+                    MutexLocker locker(queue->lock);
+                    queue->dump(out);
+                }
+                out.print("> ");
+            }
+        }
+        out.print("\n--------------------------------------------------\n");
+    }
+
+private:
+    struct RegisteredVMs {
+        WTF_MAKE_FAST_ALLOCATED;
+
+    public:
+        void add(VM* vm)
+        {
+            MutexLocker locker(lock);
+            set.add(vm);
+            if (WaiterListsManagerInternal::verbose)
+                dataLogF("WaiterListsManager registered vm %p\n", vm);
+        }
+
+        void remove(VM* vm)
+        {
+            MutexLocker locker(lock);
+            set.remove(vm);
+            if (WaiterListsManagerInternal::verbose)
+                dataLogF("WaiterListsManager unregistered vm %p\n", vm);
+        }
+
+        bool contains(VM* vm)
+        {
+            MutexLocker locker(lock);
+            return set.contains(vm);
+        }
+
+        void dump(PrintStream& out) const
+        {
+            MutexLocker locker(lock);
+            out.print("m_registeredVMs (size ", set.size(), "): ");
+            for (auto& vm : set)
+                out.print(RawPointer(vm), " ");
+        }
+
+    private:
+        mutable Mutex lock;
+        HashSet<VM*> set;
+    };
+
+    bool notifyWaiterImpl(void* ptr, std::unique_ptr<Waiter> waiter, bool isOKString)
+    {
+        if (!waiter || !m_registeredVMs.contains(waiter->getVM()))
+            return false;
+
+        if (waiter->isAsync()) {
+            VM* vm = waiter->getVM();
+            auto ticket = waiter->getTicket();
+            vm->deferredWorkTimer->scheduleWorkSoon(ticket, [isOKString](DeferredWorkTimer::Ticket ticket) {
+                JSPromise* promise = jsCast<JSPromise*>(ticket->target());
+                JSGlobalObject* globalObject = promise->globalObject();
+                VM& vm = promise->globalObject()->vm();
+                promise->resolve(globalObject, isOKString ? vm.smallStrings.okString() : vm.smallStrings.timedOutString());
+            });
+            return true;
+        }
+        return ParkingLot::unparkCount(ptr, 1);
+    }
```

Comment:
This looks not OK since non-async case is locked by a global lock.
I think we should keep non-async notification the same to the old one. So we should have appropriate mechanism to avoid a global locking for the non-async notification.
Can you explore the implementation of `ParkingLot` etc. to find the right way to do these things?

Reply:
What do you mean `global lock`? Is it `WaiterListManager.lock`? `notifyWaiterImpl` is only locked by `WaiterList.lock` which is required and specified by the spec (see `Atomics.notify` in https://tc39.es/ecma262/#sec-atomics.notify). This is because `WaiterList` is shared by both `Waiter` and `AsyncWaiter`.

```
9. Perform EnterCriticalSection(WL).
10. Let S be RemoveWaiters(WL, c).
11. Repeat, while S is not empty,
    a. Let W be the first agent in S.
    b. Remove W from the front of S.
    c. Perform NotifyWaiter(WL, W).
    d. Set n to n + 1.
12. Perform LeaveCriticalSection(WL).
```

Reply:
In `bool dequeue(const void* address, BucketMode bucketMode, const DequeueFunctor& dequeueFunctor,
    const FinishFunctor& finishFunctor)`of `ParkingLot.cpp`, it uses `Atomic.load` to get the `Bucket` (`queue`) for the given `address` (`ptr`).